packages/ens-utils/src/ethregistrar.ts

```diff
@@ -3,323 +3,370 @@ import {
+    if (name.labels.length !== 2) return null;
+
+    // last label must be "eth"
+    if (name.labels[1] !== ETH_TLD) return null;
+
+    // NOTE: now we know we have a direct subname of ".eth"
+
+    // first label must be of sufficient length
+    const subnameLength = charCount(name.labels[0]);
+    if (subnameLength < MIN_ETH_REGISTRABLE_LABEL_LENGTH) return null;
+
+    // TODO: also add a check for a maximum length limit as enforced by max block size, etc?
+
+    return buildENSName(name.labels[0]);
+  };
+
+  public getValidatedSubname = (name: ENSName): ENSName => {
+    const subname = this.getManagedSubname(name);
+    if (subname === null)
+      throw new RegistrarUnsupportedNameError(
+        "Name is not directly managed by the .ETH registrar",
+        name,
       );
-    }
-    return buildTokenId(BigInt(namehash(name.name)));
+
+    return subname;
+  };
+
+  public getContractRef(): ContractRef {
+    return this.registrarContract;
   }
 
-  isClaimable(name: ENSName, isWrapped: boolean): boolean {
-    if (!isWrapped) return false;
+  public getOnchainRegistry(): ContractRef {
+    return MAINNET_ENS_REGISTRY_WITH_FALLBACK_CONTRACT;
+  }
+
+  public canRegister(
+    name: ENSName,
+    atTimestamp: Timestamp,
+    duration: Duration,
+    existingRegistration?: Registration,
+  ): boolean {
+    if (!this.getManagedSubname(name)) {
+      // name is not directly managed by this registrar
+      return false;
+    }
 
-    if (name.labels.length >= 2) {
-      if (name.labels[1] === ETH_TLD) {
-        // first label must be of sufficient length
-        if (charCount(name.labels[0]) < MIN_ETH_REGISTRABLE_LABEL_LENGTH)
-          return false;
+    if (existingRegistration) {
+      const releaseTimestamp = getDomainReleaseTimestamp(existingRegistration);
+
+      if (releaseTimestamp && releaseTimestamp.time > atTimestamp.time) {
+        // if the name is not yet released, we can't register it
+        // TODO: check for possible off-by-1 errors in the logic above
+        return false;
       }
     }
 
-    // TODO: refine this. For example, there's a maximum length limit, etc.
-    return true;
-  }
-}
+    if (duration.seconds < 1n) {
+      // TODO: enforce that `duration` is the minimum duration allowed for a registration
+      // TODO: need to put the right constant here.
+      return false;
+    }
 
-export class ClassicETHRegistrarController implements Registrar {
-  constructor(public contract: ContractRef) {
-    this.contract = contract;
+    return true;
   }
 
-  getTokenId(name: ENSName, isWrapped: boolean): TokenId {
-    if (!this.isClaimable(name, isWrapped)) {
+  public getRegistrationPriceQuote(
+    name: ENSName,
+    atTimestamp: Timestamp,
+    duration: Duration,
+    existingRegistration?: Registration,
+  ): RegistrationPriceQuote {
+    if (!this.canRegister(name, atTimestamp, duration, existingRegistration)) {
+      // TODO: refine the way we handle these exception cases.
```

Comment:
Do you mean creating specific error handling for each use case?

e.g.

Error 1: `atTimestamp` cannot be in the past
Error 2: `duration` cannot be 0 seconds

And so on, creating better described errors... OR do you suggest something diff?

Reply:
@FrancoAguzzi I see you have a few questions about the possibility of `atTimestamp` being in the past. It's important to adjust the way you're thinking about the goals this class works to achieve.

It's totally outside the scope of concerns for this class if the params you pass into it are aligned with reality or not. Instead, this class models the behaviour of the .eth registrar contracts assuming that the inputs you provide to it were actually reality.